docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,601 @@
+
+The immediate benefit is that error handling is done for you, and you don’t have to implement any specific logic in the service itself. We’ll see later how to handle errors.
+
+Let’s see what steps are available and how to use them.
+
+### `step`
+
+This is the generic step, you provide a name, and it will run the defined method of the same name. The return value of this step doesn’t impact the execution flow, to mark the service as failed in a generic step, you have to call `#fail!` explicitly.
+
+### `model`
+
+This specialized step helps to remove some boilerplate when dealing with models. By default, it will execute the method named `fetch_<name>`. In the above example, you can see we name our model `:user` and the corresponding method is named `fetch_user`.
+
+Here, you can fetch (or instantiate) a model as you see fit. If the step returns a falsy value, then the execution flow will stop here. If an `ActiveRecord` model is returned, it will call `#invalid?` on it to determine whether the model is valid. If not, the execution flow will stop.
+
+This step is also compatible with collections: if a collection is fetched but empty, the execution flow will stop.
+
+Finally, you can also provide an `ActiveRecord` relation. The step won’t load records but will determine if the relation will return any records. If there are none, the execution flow will stop.
+
+### `policy`
+
+This step will execute the method of the same name. You can put arbitrary code here, and the execution flow will stop if the return value is falsy.
+
+Usually, a policy is related to some state on one of the service models and/or to the current user (if any).
+
+### `contract`
+
+This is one of the most powerful steps. Its main purpose is to validate the incoming data before feeding it to the models and to the service at a more global level. This is actually `ActiveModel` validations but applied to the incoming parameters.
+
+By default, this step will run coercions and validations defined in the `Contract` class. If the contract isn’t valid (at least one validation failed), then the execution flow will stop.
+
+### `transaction`
+
+This step is a bit special, as it will wrap any other steps defined in its block inside a SQL transaction.
+
+### Steps arguments
+
+Each step is called with the service context. To access a value in it, just provide its key as a keyword argument.
+
+## The context object
+
+The only state a service maintain is its context object. This is where each step can put data to be used by other steps. Most of the time, you don’t need to access the context directly, as specialized steps will store the proper data for you. But sometimes it’s necessary. In those cases, it’s just a matter of using the context like you would with a hash:
```

Comment:
I was originally thrown off by this sentence, because in the opening we say "A service object is stateless..."

I wonder if it'd help to change this to something like:

"While a service object holds no state between executions, during an execution all state between steps is maintained in its context object."

Reply:
Updated too!